code_generator/GdkCodeGenerator/Templates/UnityReaderWriterGenerator.tt

```diff
@@ -0,0 +1,71 @@
+<#@ template language="C#" #>
+<#@ output extension=".cs" #>
+<#
+    var generatedHeader = CommonGeneratorUtils.GetGeneratedHeader();
+    var componentDetails = GetComponentDetails();
+    var fieldDetailsList = GetFieldDetailsList();
+    var eventDetailsList = GetEventDetailsList();
+    var commandDetailsList = GetCommandDetailsList();
+#>
+<#= generatedHeader #>
+
+using Unity.Entities;
+using Improbable.Gdk.Core;
+using Improbable.Gdk.Core.MonoBehaviours;
+using Improbable.Worker;
+using Entity = Unity.Entities.Entity;
+
+namespace <#= qualifiedNamespace #>
+{
+    public partial class <#= componentDetails.ComponentName #>
+    {
+        [ComponentId(<#= unityComponentDefinition.Id #>)]
+        internal class ReaderWriterCreator : IReaderWriterCreator
+        {
+            public IReaderInternal CreateReaderWriter(Entity entity, EntityManager entityManager, ILogDispatcher logDispatcher)
+            {
+                return new ReaderWriterImpl(entity, entityManager, logDispatcher);
+            }
+        }
+
+        [ReaderInterface]
+        [ComponentId(<#= unityComponentDefinition.Id #>)]
+        public interface Reader : IReader<<#= componentDetails.TypeName #>>
+        {
+        }
+
+        public class ReaderWriterImpl : Reader, IReaderInternal
+        {
+            public ReaderWriterImpl(Entity entity, EntityManager entityManager, ILogDispatcher logDispatcher)
+            {
+            }
+
+            // IReader methods
+            public Authority Authority { get; }
+            public <#= componentDetails.TypeName #> Data { get; }
+            public event GameObjectDelegates.AuthorityChanged AuthorityChanged;
+            public event GameObjectDelegates.ComponentUpdated<<#= componentDetails.TypeName #>> ComponentUpdated;
+
+            // IReaderInternal methods
+            void IReaderInternal.OnAuthorityChange(Authority authority)
```

Comment:
I'm still a little unsure why IReaderInternal is used. Is it just to stop us returning object from functions. You need to know the concrete type to call any of these methods.

Reply:
Yes, technically the factory method can just return an object since all we need to reflectively inject a field is an object instance. We are returning an IReaderInternal because of a comment earlier in the PR but don't need to invoke any IReaderInternal specific methods on the object.

Reply:
In which case why does IReaderInternal have methods on it. If it's just to mark it as a specific type. I guess it adds a contract but it's a little odd to have methods on an interface that will never be called by that interface.

Reply:
The IReaderInternal methods will be called elsewhere. They are currently returned for the sake of returning some parenttype in order to not return objects (which is also fine).

Reply:
My point is that you need the concrete type to call the methods. So why not just define them on the type if this is in codegen.

Reply:
Discussed offline. @samiwh made good points about whether it makes sense to ditch the Reader, Writer interfaces altogether and just have classes with concrete implementation. This shall be taken into consideration when we flesh out the actual readerwriter code. cc @firtina-improbable